raphtory-api/src/core/utils/time.rs

```diff
@@ -0,0 +1,209 @@
+        TimeIndexEntry::from(self.and_utc().timestamp_millis())
+    }
+}
+
+impl IntoTime for TimeIndexEntry {
+    fn into_time(self) -> TimeIndexEntry {
+        self
+    }
+}
+
+impl<T: IntoTime> IntoTime for (T, usize) {
+    fn into_time(self) -> TimeIndexEntry {
+        self.0.into_time().set_index(self.1)
+    }
+}
+
+pub trait TryIntoTime {
+    fn try_into_time(self) -> Result<TimeIndexEntry, ParseTimeError>;
+}
+
+impl<T: IntoTime> TryIntoTime for T {
+    fn try_into_time(self) -> Result<TimeIndexEntry, ParseTimeError> {
+        Ok(TimeIndexEntry::from(self.into_time()))
+    }
+}
+
+impl TryIntoTime for &str {
+    /// Tries to parse the timestamp as RFC3339 and then as ISO 8601 with local format and all
+    /// fields mandatory except for milliseconds and allows replacing the T with a space
+    fn try_into_time(self) -> Result<TimeIndexEntry, ParseTimeError> {
+        let rfc_result = DateTime::parse_from_rfc3339(self);
+        if let Ok(datetime) = rfc_result {
+            return Ok(TimeIndexEntry::from(datetime.timestamp_millis()));
+        }
+
+        let result = DateTime::parse_from_rfc2822(self);
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.timestamp_millis()));
+        }
+
+        let result = NaiveDate::parse_from_str(self, "%Y-%m-%d");
+        if let Ok(date) = result {
+            let timestamp = date
+                .and_hms_opt(00, 00, 00)
+                .unwrap()
+                .and_utc()
+                .timestamp_millis();
+            return Ok(TimeIndexEntry::from(timestamp));
+        }
+
+        let result = NaiveDateTime::parse_from_str(self, "%Y-%m-%dT%H:%M:%S%.3f");
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.and_utc().timestamp_millis()));
+        }
+
+        let result = NaiveDateTime::parse_from_str(self, "%Y-%m-%dT%H:%M:%S%");
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.and_utc().timestamp_millis()));
+        }
+
+        let result = NaiveDateTime::parse_from_str(self, "%Y-%m-%d %H:%M:%S%.3f");
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.and_utc().timestamp_millis()));
+        }
+
+        let result = NaiveDateTime::parse_from_str(self, "%Y-%m-%d %H:%M:%S%");
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.and_utc().timestamp_millis()));
+        }
+
+        Err(ParseTimeError::InvalidDateTimeString(self.to_string()))
+    }
+}
+
+pub trait TryIntoTimeNeedsSecondaryIndex: TryIntoTime {}
+
+impl TryIntoTimeNeedsSecondaryIndex for i64 {}
+
+impl<Tz: TimeZone> TryIntoTimeNeedsSecondaryIndex for DateTime<Tz> {}
+
+impl TryIntoTimeNeedsSecondaryIndex for NaiveDateTime {}
+
+impl TryIntoTimeNeedsSecondaryIndex for &str {}
+
+/// Used to handle automatic injection of secondary index if not explicitly provided
+pub enum InputTime {
```

Comment:
you probably want to implement FromPyObject for this to handle the `at`, `after` problem